File: ESEOmega/Sponsors/SponsorsTVC.m
//
//  SponsorsTVC.m
//  ESEOmega
//
//  Created by Thomas Naudet on 22/07/2015.
//  Copyright © 2015 Thomas Naudet

//  This program is free software: you can redistribute it and/or modify
//  it under the terms of the GNU General Public License as published by
//  the Free Software Foundation, either version 3 of the License, or
//  (at your option) any later version.

//  This program is distributed in the hope that it will be useful,
//  but WITHOUT ANY WARRANTY; without even the implied warranty of
//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//  GNU General Public License for more details.

//  You should have received a copy of the GNU General Public License
//  along with this program. If not, see http://www.gnu.org/licenses/
//

#import "SponsorsTVC.h"

@implementation SponsorsTVC

- (void) viewDidLoad
{
    [super viewDidLoad];
    
    self.tableView.emptyDataSetSource = self;
    self.tableView.emptyDataSetDelegate = self;
    if (SYSTEM_VERSION_GREATERTHAN_OR_EQUALTO(@"10.0")) {
        self.tableView.prefetchDataSource = self;
    }
    self.tableView.rowHeight = UITableViewAutomaticDimension;
    self.tableView.estimatedRowHeight = 131.;
    self.refreshControl.tintColor = [UINavigationBar appearance].barTintColor;
    
    if (self.traitCollection.forceTouchCapability == UIForceTouchCapabilityAvailable)
        [self registerForPreviewingWithDelegate:self sourceView:self.tableView];
    
    NSNotificationCenter *ctr = [NSNotificationCenter defaultCenter];
    [ctr addObserver:self selector:@selector(loadSponsors) name:@"sponsors" object:nil];
    [ctr addObserver:self.refreshControl selector:@selector(endRefreshing) name:@"debugRefresh" object:nil];
    [ctr addObserver:self.refreshControl selector:@selector(endRefreshing) name:@"sponsorsSent" object:nil];
    
    [self.tableView setBackgroundColor:[UIColor groupTableViewBackgroundColor]];
    self.tableView.tableFooterView = [UIView new];
    [self loadSponsors];
}

- (void) viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:animated];
    [self.refreshControl endRefreshing];
    
    // Handoff
    NSUserActivity *activity = [[NSUserActivity alloc] initWithActivityType:@"com.eseomega.ESEOmega.sponsors"];
    activity.title = @"Partenaires BDE ESEO";
    activity.webpageURL = [NSURL URLWithString:URL_ACT_SPON];
    activity.eligibleForSearch = YES;
    activity.eligibleForHandoff = YES;
    activity.eligibleForPublicIndexing = YES;
    self.userActivity = activity;
    [self.userActivity becomeCurrent];
}

- (void) dealloc
{
    [[NSNotificationCenter defaultCenter] removeObserver:self];
}

- (UIStatusBarStyle) preferredStatusBarStyle
{
    return UIStatusBarStyleLightContent;
}

#pragma mark - Actions

- (void) recupSponsors:(BOOL)forcer
{
    if (!forcer && ![[Data sharedData] shouldUpdateJSON:@"sponsors"])
    {
        [self.refreshControl endRefreshing];
        return;
    }
    
    [[Data sharedData] updateJSON:@"sponsors"];
}

- (void) loadSponsors
{
    sponsors = [[Data sharedData] sponsors][@"sponsors"];
    
    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
        
        NSMutableArray *t_sponsorsDescAttributed = [NSMutableArray arrayWithCapacity:sponsors.count];
        
        for (NSDictionary *sponsor in sponsors)
        {
            /* Description label, allow HTML */
            NSString *detail = [sponsor[@"description"] stringByReplacingOccurrencesOfString:@"\\n" withString:@"\n"];
            detail = [detail stringByReplacingOccurrencesOfString:@" €" withString:@" €"];  // no-breaking space
            detail = [detail stringByReplacingOccurrencesOfString:@" %" withString:@" %"];  // nbsp
            detail = [[NSString stringWithFormat:@"<style>body{font-family: '%@'; font-size:%dpx;}</style>",
                       [UIFont systemFontOfSize:15].fontName, 15]
                      stringByAppendingString:detail];
            // Remove trailing HTML useless tags
            while ([detail hasSuffix:@"<br>"] || [detail hasSuffix:@"<br/>"] || [detail hasSuffix:@"<br />"]) {
                detail = [detail substringToIndex:[detail rangeOfString:@"<br" options:NSBackwardsSearch].location];
            }
            NSAttributedString *as = [[NSAttributedString alloc] initWithData:[detail dataUsingEncoding:NSUnicodeStringEncoding]
                                                                      options:@{ NSDocumentTypeDocumentAttribute: NSHTMLTextDocumentType }
                                                           documentAttributes:nil error:nil];
            [t_sponsorsDescAttributed addObject:as];
        }
        
        sponsorsDescAttributed = t_sponsorsDescAttributed;
        
        dispatch_async(dispatch_get_main_queue(), ^(void){
            
            CATransition *animation = [CATransition animation];
            [animation setDuration:0.2f];
            [animation setTimingFunction:[CAMediaTimingFunction functionWithName:kCAMediaTimingFunctionEaseOut]];
            [animation setType:kCATransitionFade];
            [self.tableView.layer addAnimation:animation forKey:NULL];
            
            [self.tableView reloadData];
            
            if ([sponsorsDescAttributed count])
            {
                [self.tableView setBackgroundColor:[UIColor whiteColor]];
                self.tableView.tableFooterView = nil;
            }
            else
            {
                [self.tableView setBackgroundColor:[UIColor groupTableViewBackgroundColor]];
                self.tableView.tableFooterView = [UIView new];
            }
        });
    });
    
    if ([sponsorsDescAttributed count])
    {
        [self.tableView setBackgroundColor:[UIColor whiteColor]];
        self.tableView.tableFooterView = nil;
    }
    else
    {
        [self.tableView setBackgroundColor:[UIColor groupTableViewBackgroundColor]];
        self.tableView.tableFooterView = [UIView new];
    }
}

- (IBAction) refresh:(UIRefreshControl *)sender
{
    [self recupSponsors:NO];
}

#pragma mark - Table view data source

- (NSInteger) numberOfSectionsInTableView:(nonnull UITableView *)tableView
{
    return 1;
}

- (NSInteger) tableView:(nonnull UITableView *)tableView
  numberOfRowsInSection:(NSInteger)section
{
    return [sponsorsDescAttributed count];
}

- (nonnull UITableViewCell *)tableView:(nonnull UITableView *)tableView
                 cellForRowAtIndexPath:(nonnull NSIndexPath *)indexPath
{
    SponsorsCell *cell = [tableView dequeueReusableCellWithIdentifier:@"sponsorsCell" forIndexPath:indexPath];
    
    /* Get current data and set name */
    NSDictionary *sponsor = sponsors[indexPath.row];
    cell.nomLabel.text = sponsor[@"name"];
    
    cell.descLabel.attributedText = sponsorsDescAttributed[indexPath.row];
    
    /* Build URL/address contact string */
    NSString *contact = @"";
    if (sponsor[@"url"] != nil) {
        /* Remove unnecessary spaces and slashes */
        NSString *url = sponsor[@"url"];
        url = [url stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceCharacterSet]];
        if ([url hasSuffix:@"/"])
            url = [url substringToIndex:url.length - 1];
        /* Remove old-school URL parts */
        contact = [[[url stringByReplacingOccurrencesOfString:@"http://"
                                                   withString:@""] stringByReplacingOccurrencesOfString:@"www." withString:@""] stringByReplacingOccurrencesOfString:@"https://" withString:@""];
    }
    if (![contact isEqualToString:@""] && sponsor[@"address"] != nil && ![sponsor[@"address"] isEqualToString:@""])
        contact = [contact stringByAppendingString:@"\n"];
    if (sponsor[@"address"] != nil)
        contact = [contact stringByAppendingString:sponsor[@"address"]];
    cell.contactLabel.text = contact;
    
    /* Set logo */
    if (sponsor[@"img"] != nil && ![sponsor[@"img"] isEqualToString:@""])
        [cell.logoView sd_setImageWithURL:[NSURL URLWithString:sponsor[@"img"]]
                         placeholderImage:[UIImage imageNamed:@"placeholder"]];
    else
        [cell.logoView setImage:[UIImage imageNamed:@"placeholder"]];
    cell.logoView.layer.cornerRadius = 4;
    cell.logoView.clipsToBounds = YES;
    
    /* Advantages */
    [cell setAvantages:sponsor[@"perks"]];
    
    /* Disable selection if no link */
    [cell setSelectionStyle:(sponsor[@"url"] == nil || [sponsor[@"url"] isEqualToString:@""]) ? UITableViewCellSelectionStyleNone
                                                                                              : UITableViewCellSelectionStyleDefault];
    
    return cell;
}

#pragma mark - Table view data source prefetching

/**
 Prepare data (sponsor logo) at specified index paths
 
 @param tableView This table view
 @param indexPath Position of the cells to preload
 */
- (void)       tableView:(UITableView *)tableView
prefetchRowsAtIndexPaths:(NSArray<NSIndexPath *> *)indexPaths
{
    /* Get every image URL to fetch */
    NSMutableArray *thumbnails = [NSMutableArray arrayWithCapacity:indexPaths.count];
    for (NSIndexPath *indexPath in indexPaths)
    {
        NSString *imgURL = sponsors[indexPath.row][@"img"];
        if (imgURL != nil && ![imgURL isEqualToString:@""])
            [thumbnails addObject:[NSURL URLWithString:imgURL]];
    }
    
    [[SDWebImagePrefetcher sharedImagePrefetcher] prefetchURLs:thumbnails];
}

#pragma mark - Table view delegate

- (void)      tableView:(nonnull UITableView *)tableView
didSelectRowAtIndexPath:(nonnull NSIndexPath *)indexPath
{
    NSString *url = sponsors[indexPath.row][@"url"];
    if (url == nil || [url isEqualToString:@""])
        return;
    
    [[Data sharedData] openURL:url currentVC:self];
    
    [tableView deselectRowAtIndexPath:indexPath animated:YES];
}

#pragma mark - 3D Touch

- (UIViewController *) previewingContext:(id<UIViewControllerPreviewing>)previewingContext
               viewControllerForLocation:(CGPoint)location
{
    NSIndexPath *index = [self.tableView indexPathForRowAtPoint:location];
    if (index != nil)
    {
        NSString *url = sponsors[index.row][@"url"];
        if (url == nil || [url isEqualToString:@""])
            return nil;
        
        previewingContext.sourceRect = [self.tableView rectForRowAtIndexPath:index];
        
        SFSafariViewController *safari = [[SFSafariViewController alloc] initWithURL:[NSURL URLWithString:[url stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceCharacterSet]]]
                                                             entersReaderIfAvailable:NO];
        if ([SFSafariViewController instancesRespondToSelector:@selector(preferredBarTintColor)])
        {
            safari.preferredBarTintColor = [UINavigationBar appearance].barTintColor;
            safari.preferredControlTintColor = [UINavigationBar appearance].tintColor;
        }
        return safari;
    }
    return nil;
}

- (void) previewingContext:(id<UIViewControllerPreviewing>)previewingContext
      commitViewController:(UIViewController *)viewControllerToCommit
{
    [self presentViewController:viewControllerToCommit animated:YES completion:nil];
}

#pragma mark - Handoff

- (void) updateUserActivityState:(NSUserActivity *)activity
{
    [super updateUserActivityState:activity];
}

#pragma mark - DZNEmptyDataSet

- (UIImage *) imageForEmptyDataSet:(UIScrollView *)scrollView
{
    return [UIImage imageNamed:@"autreVide"];
}

- (NSAttributedString *) titleForEmptyDataSet:(UIScrollView *)scrollView
{
    NSString *text = @"Aucun partenaire";
    
    NSDictionary *attributes = @{NSFontAttributeName: [UIFont boldSystemFontOfSize:18.0],
                                 NSForegroundColorAttributeName: [UIColor darkGrayColor]};
    
    return [[NSAttributedString alloc] initWithString:text attributes:attributes];
}

- (NSAttributedString *) descriptionForEmptyDataSet:(UIScrollView *)scrollView
{
    NSString *text = @"Vérifiez votre connexion et tirez pour rafraîchir.";
    
    NSMutableParagraphStyle *paragraph = [NSMutableParagraphStyle new];
    paragraph.lineBreakMode = NSLineBreakByWordWrapping;
    paragraph.alignment = NSTextAlignmentCenter;
    
    NSDictionary *attributes = @{NSFontAttributeName: [UIFont systemFontOfSize:14.0],
                                 NSForegroundColorAttributeName: [UIColor lightGrayColor],
                                 NSParagraphStyleAttributeName: paragraph};
    
    return [[NSAttributedString alloc] initWithString:text attributes:attributes];
}

- (CGPoint) offsetForEmptyDataSet:(UIScrollView *)scrollView
{
    return CGPointMake(0, -self.tableView.tableHeaderView.frame.size.height / 2. - 25.5);
}

- (UIColor *) backgroundColorForEmptyDataSet:(UIScrollView *)scrollView
{
    return [UIColor groupTableViewBackgroundColor];
}

@end

File: ESEOmega/Sponsors/SponsorsTVC.swift
//
//  SponsorsTVCTableViewController.swift
//  BDE-ESEO
//
//  Created by Benjamin Gondange on 10/09/2018.
//  Copyright © 2018 Benjamin Gondange

//  This program is free software: you can redistribute it and/or modify
//  it under the terms of the GNU General Public License as published by
//  the Free Software Foundation, either version 3 of the License, or
//  (at your option) any later version.

//  This program is distributed in the hope that it will be useful,
//  but WITHOUT ANY WARRANTY; without even the implied warranty of
//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//  GNU General Public License for more details.

//  You should have received a copy of the GNU General Public License
//  along with this program. If not, see http://www.gnu.org/licenses/
//

import UIKit
import SafariServices

class SponsorsTVC: UITableViewController {
    private let reuseIdentifier     = "sponsorsCell"
    
    var sponsors = [Sponsor]()
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        loadFromCache()
        fetchRemote()
        
        if #available(iOS 10.0, *) {
            self.tableView.prefetchDataSource = self;
        }
        
        self.tableView.rowHeight = UITableView.automaticDimension;
        self.tableView.estimatedRowHeight = 131;
        self.refreshControl?.tintColor = UINavigationBar.appearance().tintColor
        
        /* Refresh control */
        reloadRefreshControl()
        
        if (self.traitCollection.forceTouchCapability == UIForceTouchCapability.available) {
            self.registerForPreviewing(with: self, sourceView: self.tableView)
        }
        
        
        
        // Uncomment the following line to preserve selection between presentations
        // self.clearsSelectionOnViewWillAppear = false

        // Uncomment the following line to display an Edit button in the navigation bar for this view controller.
        // self.navigationItem.rightBarButtonItem = self.editButtonItem
    }

    override func viewWillAppear(_ animated: Bool) {
        super .viewWillAppear(animated)
        self.refreshControl?.endRefreshing()
        
        let activity = NSUserActivity(activityType: "com.eseomega.ESEOmega.sponsors")
        activity.title = "Liste des partenaires"
        activity.webpageURL = URL(string: URL_ACT_SPON)
        activity.isEligibleForSearch = true
        activity.isEligibleForHandoff = true
        activity.isEligibleForPublicIndexing = true
        self.userActivity = activity
        self.userActivity?.becomeCurrent()
    }
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }

    func reloadRefreshControl() {
        
        let refreshText = "Charger les sponsors récents…"
        var refreshColor: UIColor = .white
        
        if #available(iOS 11.0, *) {
        } else {
            refreshColor = UINavigationBar.appearance().barTintColor ?? .blue
        }
        
        refreshControl?.tintColor = refreshColor
        refreshControl?.attributedTitle = NSAttributedString(string: refreshText,
                                                             attributes: [.foregroundColor : refreshColor])
    }
    @IBAction func refresh(_ sender: Any) {
        self.fetchRemote()
    }
    
}

extension SponsorsTVC: APIViewer {
    typealias T = [Sponsor]
    
    func loadFromCache() {
        let decoder = JSONDecoder()
        guard let data   = APIArchiver.getCache(for: .sponsors),
            let result = try? decoder.decode([Sponsor].self, from: data) else {
                reloadData()
                return
        }
        
        self.loadData(result)
    }
    
    func fetchRemote() {
        
        API.request(.sponsors, get: [:],
                    completed: { data in
                        
                        DispatchQueue.main.async {
                            self.refreshControl?.endRefreshing()
                        }
                        
                        let decoder = JSONDecoder()
                        
                        guard let result = try? decoder.decode(SponsorsResult.self, from: data),
                            result.success
                            else { return }
                        self.loadData(result.sponsors)
                        APIArchiver.save(data: result.sponsors, for: .sponsors)
                        
        }, failure: { _, _ in
            DispatchQueue.main.async {
                self.refreshControl?.endRefreshing()
            }
        })
    }
    func loadData(_ data: [Sponsor]) {
        
        let atLaunch = sponsors.isEmpty
        /* Add new articles.
         Replace the same old articles with fresh content. */
        if atLaunch {
            sponsors = data
        }
        
        self.reloadData()
    }
    func reloadData() {
        
        DispatchQueue.main.async {
            if self.sponsors.isEmpty {
                self.tableView.backgroundColor = .groupTableViewBackground
                self.tableView.tableFooterView = UITableViewHeaderFooterView()
            } else {
                self.tableView.backgroundColor = .white
                self.tableView.tableFooterView = nil
            }
            self.tableView.reloadData()
        }
    }
   
}

extension SponsorsTVC {
    override func tableView(_ tableView: UITableView,
                            numberOfRowsInSection section: Int) -> Int {
        
        return sponsors.isEmpty ? 0 : (sponsors.count + 1)
    }
    
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        guard indexPath.row != sponsors.count else {
            return UITableViewCell()
        }
        
        let cell = tableView.dequeueReusableCell(withIdentifier: reuseIdentifier, for: indexPath) as! SponsorsCell
        
        let sponsor = sponsors[indexPath.row]
        
        cell.nomLabel?.text = sponsor.name
        cell.descLabel?.text = sponsor.description
        
        var contact = ""
        if (sponsor.url != nil && sponsor.url != "") {
            guard var url = sponsor.url else { return UITableViewCell() }
            NSLog(url)
            url = url.trimmingCharacters(in: CharacterSet.whitespaces)
            if (url.hasSuffix("/")) {
                let index = url.index(url.startIndex, offsetBy: url.count - 1)
                url = String(url.prefix(upTo: index))
            }
            contact = url.replacingOccurrences(of: "https://", with: "").replacingOccurrences(of: "www.", with: "").replacingOccurrences(of: "http://", with: "")
            
        }
        
        if (contact != "" && sponsor.address != nil && sponsor.address != "") {
            contact = contact + "\n"
        }
        
        guard let add = sponsor.address else { return UITableViewCell() }
        contact = contact + add
        
        cell.contactLabel?.text = contact
        
        if let image = sponsor.image {
            cell.logoView?.sd_setImage(with: URL(string: image), placeholderImage: UIImage(named: "placeholder"))
        } else {
            cell.logoView?.image = UIImage(named: "placeholder")
        }
        cell.logoView?.layer.cornerRadius = 4
        cell.logoView?.clipsToBounds = true
        
        cell.setAvantages(avt: sponsor.perks)
        
        return cell
    }
}

// MARK: - Table View Data Source Prefetching
extension SponsorsTVC: UITableViewDataSourcePrefetching {
    
    /// Prepares data (news image) at specified index paths
    ///
    /// - Parameters:
    ///   - tableView: This table view
    ///   - indexPaths: Position of the cells to preload
    func tableView(_ tableView: UITableView,
                   prefetchRowsAt indexPaths: [IndexPath]) {
        
        let thumbnailURLs: [URL] = indexPaths.compactMap { indexPath in
            
            guard indexPath.row != sponsors.count,
                let articleImg = sponsors[indexPath.row].image
                else { return nil }
            
            return URL(string: articleImg)
        }
        SDWebImagePrefetcher.shared().prefetchURLs(thumbnailURLs)
    }
    
}
// MARK: - Table view controller Delegate
extension SponsorsTVC {
    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        guard let url = URL(string: sponsors[indexPath.row].url ?? "") else {
            self.tableView.deselectRow(at: indexPath, animated: true)
            return
            
        }

        let svc = SFSafariViewController(url: url)
        self.present(svc, animated: true, completion: nil)
        
        self.tableView.deselectRow(at: indexPath, animated: true)
    }
}

extension SponsorsTVC: UIViewControllerPreviewingDelegate {
    
    func previewingContext(_ previewingContext: UIViewControllerPreviewing,
                           viewControllerForLocation location: CGPoint) -> UIViewController? {
        
        // TODO: - Preview delegate
        return nil
    }
    
    func previewingContext(_ previewingContext: UIViewControllerPreviewing,
                           commit viewControllerToCommit: UIViewController) {
        
    }
    
}

// MARK: - Empty Data Set Data Source
extension SponsorsTVC: DZNEmptyDataSetSource {
    
    func image(forEmptyDataSet scrollView: UIScrollView!) -> UIImage! {
        
        return #imageLiteral(resourceName: "newsVide")
    }
    
    func title(forEmptyDataSet scrollView: UIScrollView!) -> NSAttributedString! {
        
        return NSAttributedString(string: "Aucun sponsor",
                                  attributes: [.foregroundColor : UIColor.darkGray])
    }
    
    func description(forEmptyDataSet scrollView: UIScrollView!) -> NSAttributedString! {
        
        let text = "Vérifiez votre connexion et tirez pour rafraîchir."
        
        return NSAttributedString(string: text,
                                  attributes: [.font : UIFont.preferredFont(forTextStyle: .subheadline),
                                               .foregroundColor : UIColor.lightGray])
    }
    
    func backgroundColor(forEmptyDataSet scrollView: UIScrollView!) -> UIColor! {
        
        return .groupTableViewBackground
    }
    
}


